### PR TITLE
Support a global log HTTP path prefix flag and custom per log path prefix

### PR DESCRIPTION
**Pull request by Martin2112**
Setting the global prefix allows all logs to appear under a common prefix e.g. `/logs`. Setting a custom prefix allows this to be defined per log. This is intended to be useful in migration scenarios where the new logs follow a set pattern but existing logs don't.

This brings the code more in line with the Google internal version of the CT server. The current behaviour will not change unless the global handlerPrefix flag is set and/or custom prefix settings are added to the log configs.

### CODE REVIEW

File: trillian/ctfe/configpb/config.proto [hunk 1]
```diff
@@ -49,6 +49,11 @@ message LogConfig {
   // which is common for both types.
   int64 log_id = 1;
   // TODO(pavelkalinnikov): Comment this.
```
Comment (daviddrysdale) [inline trillian/ctfe/configpb/config.proto]: Can remove the TODO now

File: trillian/ctfe/configpb/config.proto [hunk 2]
```diff
@@ -88,6 +93,13 @@ message LogConfig {
   // log's key and can't sign STHs. Consequently, the log operator must ensure
   // to channel source log's STHs into CTFE.
   bool is_mirror = 12;
+  // custom_handler_prefix if set to a non empty value overrides the global
+  // handler prefix for an individual log. For example this field is set to
+  // "/otherlogs" then a log with prefix "vogon" will make it's get-sth handler
+  // available at "/otherlogs/vogon/ct/v1/get-sth" regardless of what the
+  // global prefix is. Can be set to '/' to make the get-sth handler register
+  // at "/vogon/ct/v1/get-sth".
+  string custom_handler_prefix = 13;
```
Comment (daviddrysdale) [inline trillian/ctfe/configpb/config.proto]: Can we make this simpler by dropping `custom_handler_prefix` and allowing `/` in the per-log prefix?  That way you could set the global prefix to `"/"` and have per-log prefixes of `"logs/vogon"`, `"otherlogs/dent"`, `"toplevel"` ?

Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Possibly. The advantage of this is that it doesn't change the config validation rules. Plus prefix is used in other places like metrics? so there could be unexpected additional impact of changing that rule.

File: trillian/ctfe/configpb/config.proto (hunk 1 above)
Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Done.

File: trillian/ctfe/configpb/config.proto (hunk 2 above)
Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Or is tree id the only thing metrics uses? Not entirely sure.

Comment (daviddrysdale) [reply trillian/ctfe/configpb/config.proto]: Ah, fair point about `prefix`.  Maybe:
 - rename `custom_handler_prefix` to `override_handler_prefix`
 - move the field up next to `prefix` (if that's allowed for .proto files)

Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Just waiting to see if the updated protoc-gen-go fixes the build error then I'll have another look at it.

Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Done.

File: trillian/ctfe/configpb/config.proto [hunk 3]
```diff
@@ -42,14 +42,27 @@ message LogConfigSet {
...
+// NEXT_ID: 14
 message LogConfig {
   // The ID of a Trillian tree that stores the log data. The tree type must be
   // LOG for regular CT logs. For mirror logs it must be either PREORDERED_LOG
   // or LOG, and can change at runtime. CTFE in mirror mode uses only read API
   // which is common for both types.
   int64 log_id = 1;
-  // TODO(pavelkalinnikov): Comment this.
+  // prefix is the name of the log. It will come after the global or
+  // custom handler prefix. For example if the handler prefix is "/logs"
```
Comment (daviddrysdale) [inline trillian/ctfe/configpb/config.proto]: s/custom/override/ ?

File: trillian/ctfe/ct_server/main.go [hunk 4]
```diff
@@ -289,12 +290,24 @@ func setupAndRegister(ctx context.Context, client trillian.TrillianLogClient, de
...
 		opts.CertificateQuotaUser = ctfe.QuotaUserForCert
 	}
+	// Full handler pattern will be of the form "/logs/yyz/ct/v1/add-chain", where "/logs" is the
+	// HandlerPrefix and "yyz" is the c.Prefix for this particular log. Use the default
+	// HandlerPrefix unless the log config overrides it. The custom prefix in
+	// the log configuration intended for use in migration scenarios where logs
+	// have an existing URL path that differs from the global one. For example
+	// if all new logs are served on "/logs/log/..." and a previously existing
+	// log is at "/log/..." this is now supported.
+	lhp := *handlerPrefix
```
Comment (daviddrysdale) [inline trillian/ctfe/ct_server/main.go]: nit: could we pass in `handlerPrefix` as a parameter so only `main()` touches flags please

File: trillian/ctfe/configpb/config.proto (hunk 3 above)
Comment (Martin2112) [reply trillian/ctfe/configpb/config.proto]: Done.

File: trillian/ctfe/ct_server/main.go (hunk 4 above)
Comment (Martin2112) [reply trillian/ctfe/ct_server/main.go]: Done.

**Comment by Martin2112** [comment]
Drat I seem to have somehow merged this minus the last commit.

**Pull request merged**